lib/grack/git.rb

```diff
@@ -0,0 +1,58 @@
+module Grack
+  class Git
+    attr_reader :repo
+    def initialize(git_path, repo_path)
+      @git_path = git_path
+      @repo = repo_path
+      @repo = nil unless valid_git_dir?
+    end
+
+    def update_server_info
+      execute(%W(update-server-info))
+    end
+
+    def command(cmd)
+      [@git_path || 'git'] + cmd
+    end
+
+    def capture(cmd)
+      IO.popen(popen_env, cmd, popen_options).read
+    end
+
+    def execute(cmd)
+      capture( command(cmd) ).chomp
+    end
+
+    def execute_with_block(cmd)
```

Comment:
I agree, I wasn't sure if you were ok with the `File::RDWR` mode to be on the normal read execute method but looking at it again, that can be easily solved. I'll have  another go at this.